File: codes/go/Array.reduce/Array.reduce.go
package main

import (
	"encoding/json"
	"fmt"
	"reflect"
)

const EMPTY_STRING = ""
const TAB = "    "

// type any interface{}
type array []any
type object map[string]any

func prettyJsonStringify(anything any) string {
	marshalledJson, err := json.MarshalIndent(anything, EMPTY_STRING, TAB)
	if err == nil {
		return string(marshalledJson)
	}

	return "undefined"
}

func prettyArrayOfPrimitives(anArray array) string {
	result := "["
	for arrayItemIndex, arrayItem := range anArray {
		switch arrayItemType := reflect.TypeOf(arrayItem).Kind(); arrayItemType {
		case reflect.String:
			result += "\"" + arrayItem.(string) + "\""
		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128:
			result += fmt.Sprint(arrayItem)
		default:
			continue
		}
		if ((arrayItemIndex + 1) != len(anArray)) {
			result = result + ", "
		}
	}
	result = result + "]"
	return result
}

func spreadSyntaxObject(parameters ...any) object {
	var newObject = make(object)
	for _, parameter := range parameters {
		parameterType := reflect.TypeOf(parameter).Kind()
		if (parameterType == reflect.Map) {
			for objectKey, objectValue := range parameter.(object) {
				newObject[objectKey] = objectValue
			}
		}
		if (parameterType == reflect.Slice) {
			for arrayItemIndex, arrayItem := range parameter.(array) {
				newObject[prettyJsonStringify(arrayItemIndex)] = arrayItem
			}
		}
	}
	return newObject
}

func spreadSyntaxArray(parameters ...any) array {
	var newArray = array{}
	for _, parameter := range parameters {
		parameterType := reflect.TypeOf(parameter).Kind()
		if (parameterType == reflect.Map) {
			if (len(parameter.(object)) == 1) {
				for _, objectValue := range parameter.(object) {
					newArray = append(newArray, objectValue)
				}
				continue
			}
			newArray = append(newArray, parameter)
			continue
		}
		if (parameterType == reflect.Slice) {
			newArray = append(newArray, parameter.(array)...)
			continue
		}
	}
	return newArray
}

func arrayReduce(callbackFunction func(any, any, int, array) any, anArray array, initialValue any) any {
	// JavaScript-like Array.reduce() function
	result := initialValue
	for arrayItemIndex, arrayItem := range anArray {
		result = callbackFunction(result, arrayItem, arrayItemIndex, anArray)
	}
	return result
}

func main() {
	fmt.Println("\n// JavaScript-like Array.reduce() in Go Slice")

	numbers := array{12, 34, 27, 23, 65, 93, 36, 87, 4, 254}
	fmt.Println("numbers:", prettyArrayOfPrimitives(numbers))

	fmt.Println("// using JavaScript-like Array.reduce() function \"arrayReduce\"")

	numbersTotal := arrayReduce(func(currentResult any, currentNumber any, _ int, _ array) any {
		return (currentResult.(int) + currentNumber.(int))
	}, numbers, 0)
	fmt.Println("total numbers:", prettyJsonStringify(numbersTotal))
	// total number: 635

	fmt.Println("\n// JavaScript-like Array.reduce() in Slice of Maps")

	products := array{
		object{
			"code":  "pasta",
			"price": 321,
		},
		object{
			"code":  "bubble_gum",
			"price": 233,
		},
		object{
			"code":  "potato_chips",
			"price": 5,
		},
		object{
			"code":  "towel",
			"price": 499,
		},
	}
	fmt.Println("products:", prettyJsonStringify(products))

	fmt.Println("// using JavaScript-like Array.reduce() function \"arrayReduce\"")

	productsGrouped := arrayReduce(func(currentResult any, currentProduct any, _ int, _ array) any {
		if (currentProduct.(object)["price"].(int) > 100) {
			return spreadSyntaxObject(currentResult, object{"expensive": spreadSyntaxArray(currentResult.(object)["expensive"], object{"currentProduct": currentProduct})})
		}
		return spreadSyntaxObject(currentResult, object{"cheap": spreadSyntaxArray(currentResult.(object)["cheap"], object{"currentProduct": currentProduct})})
	}, products, object{"expensive": array{}, "cheap": array{}})
	fmt.Println("grouped products:", prettyJsonStringify(productsGrouped))
	// grouped products: {
	// 	"expensive": [
	// 		{
	// 			"code": "pasta",
	// 			"price": 321
	// 		},
	// 		{
	// 			"code": "bubble_gum",
	// 			"price": 233
	// 		},
	// 		{
	// 			"code": "towel",
	// 			"price": 499
	// 		}
	// 	],
	// 	"cheap": [
	// 		{
	// 			"code": "potato_chips",
	// 			"price": 5
	// 		}
	// 	]
	// }
}

File: codes_trial/swift/Array.reduce.swift
import Foundation

typealias MyObject = [String: Any?]
typealias MyArray = [Any?]

func prettyArrayOfPrimitives(_ anArrayOfPrimitives: MyArray) -> String {
    var result = "["
    for (arrayItemIndex, arrayItem) in anArrayOfPrimitives.enumerated() {
        if (((arrayItem is String) == false) && ((arrayItem is NSNumber) == false)) {
            continue
        }
        if let arrayItem = arrayItem as? String {
            result += "\"\(arrayItem)\""
        }
        if let arrayItem = arrayItem as? NSNumber {
            result += "\(arrayItem)"
        }
        if ((arrayItemIndex + 1) != anArrayOfPrimitives.count) {
            result += ", "
        }
        
    }
    result += "]"
    return result
}

func prettyJsonStringify(_ anything: Any? = nil, indent: String? = "    ") -> String {
    var indentLevel = 0
    func prettyJsonStringifyInnerFunction(_ anything: Any? = nil, _ indent: String? = "    ") -> String {
        guard let anything = anything else {
            return "undefined"
        }
        if (anything as? String == "null") {
            return "null"
        }
        if (anything as? String == "undefined") {
            return "undefined"
        }
        if let anything = anything as? String {
            return "\"\(anything)\""
        }
        if let anything = anything as? NSNumber {
            return "\(anything)"
        }
        if let anything = anything as? Bool {
            return "\(anything)"
        }
        if let anything = anything as? MyArray {
            indentLevel += 1
            var result = "[\n\(String(repeating: indent ?? "    ", count: indentLevel))"
            for (arrayItemIndex, arrayItem) in anything.enumerated() {
                result += prettyJsonStringifyInnerFunction(arrayItem, indent)
                if ((arrayItemIndex + 1) != anything.count) {
                    result += ",\n\(String(repeating: indent ?? "    ", count: indentLevel))"
                }
            }
            indentLevel -= 1
            result += "\n\(String(repeating: indent ?? "    ", count: indentLevel))]"
            return result
        }
        if let anything = anything as? MyObject {
            indentLevel += 1
            var result = "{\n\(String(repeating: indent ?? "    ", count: indentLevel))"
            for (entryIndex, (objectKey, objectValue)) in anything.enumerated() {
                result += "\"\(objectKey)\": \(prettyJsonStringifyInnerFunction(objectValue, indent))"
                if ((entryIndex + 1) != anything.count) {
                    result += ",\n\(String(repeating: indent ?? "    ", count: indentLevel))"
                }
            }
            indentLevel -= 1
            result += "\n\(String(repeating: indent ?? "    ", count: indentLevel))}"
            return result
        }
        return "undefined"
    }
    return prettyJsonStringifyInnerFunction(anything, indent)
}

func spreadSyntaxObject(_ parameters: Any?...) -> MyObject {
    var newObject = MyObject()
    for (_, parameter) in parameters.enumerated() {
        if parameter is MyObject {
            guard let parameter = parameter as? MyObject else {
                continue
            }
            for (objectKey, objectValue) in parameter {
                newObject[objectKey] = objectValue
            }
            continue
        }
        if parameter is MyArray {
            guard let parameter = parameter as? MyArray else {
                continue
            }
            for (arrayItemIndex, arrayItem) in parameter.enumerated() {
                newObject[String(arrayItemIndex)] = arrayItem
            }
            continue
        }
    }
    return newObject
}

func spreadSyntaxArray(_ parameters: Any?...) -> MyArray {
    var newArray = MyArray()
    for (_, parameter) in parameters.enumerated() {
        if parameter is MyObject {
            guard let parameter = parameter as? MyObject else {
                continue
            }
            if (parameter.count == 1) {
                for (_, objectValue) in parameter {
                    newArray.append(objectValue)
                }
                continue
            }
            newArray.append(parameter)
            continue
        }
        if parameter is MyArray {
            guard let parameter = parameter as? MyArray else {
                continue
            }
            for (_, arrayItem) in parameter.enumerated() {
                newArray.append(arrayItem)
            }
            continue
        }
    }
    return newArray
}

func arrayReduce(_ callbackFunction: (Any?, Any?, Int, MyArray) -> Any?, _ anArray: MyArray, _ initialValue: Any?) -> Any? {
    // JavaScript-like Array.reduce() function
    var result = initialValue
    for (arrayItemIndex, arrayItem) in anArray.enumerated() {
        result = callbackFunction(result, arrayItem, arrayItemIndex, anArray)
    }
    return result
}

print("\n// JavaScript-like Array.reduce() in Swift [Any?]")

let numbers: MyArray = [12, 34, 27, 23, 65, 93, 36, 87, 4, 254]
print("numbers: \(prettyArrayOfPrimitives(numbers))")

var numbersTotal: Any?

print("// using JavaScript-like Array.reduce() function \"arrayReduce\"")

numbersTotal = arrayReduce({ (currentResult: Any?, currentNumber: Any?, _: Int, _: MyArray) -> Any? in
    guard let currentResult = currentResult as? Int, let currentNumber = currentNumber as? Int else {
        return nil
    }
    return currentResult + currentNumber
}, numbers, 0)
print("total number: \(numbersTotal ?? "undefined")")
// total number: 635

print("// using Swift Array.reduce() built-in method \"Array.reduce\"")

numbersTotal = numbers.reduce(0) { (currentResult: Any?, currentNumber: Any?) -> Any? in
    guard let currentResult = currentResult as? Int, let currentNumber = currentNumber as? Int else {
        return nil
    }
    return currentResult + currentNumber
}
print("total number: \(numbersTotal ?? "undefined")")
// total number: 635

print("\n// JavaScript-like Array.map() in Swift [[String, Any?]]")

let products: MyArray = [
    [
        "code": "pasta",
        "price": 321
    ] as MyObject,
    [
        "code": "bubble_gum",
        "price": 233
    ] as MyObject,
    [
        "code": "potato_chips",
        "price": 5
    ] as MyObject,
    [
        "code": "towel",
        "price": 499
    ] as MyObject
]
print("products: \(prettyJsonStringify(products))")

var productsGrouped: Any?

print("// using JavaScript-like Array.reduce() function \"arrayReduce\"")

productsGrouped = arrayReduce({ (currentResult: Any?, currentProduct: Any?, _: Int, _: MyArray) -> Any? in
    guard let currentResult = currentResult as? MyObject, let arrayExpensive = currentResult["expensive"], let arrayCheap = currentResult["cheap"], let currentProduct = currentProduct as? MyObject, let currentProductPrice = currentProduct["price"] as? Int else {
        return nil
    }
    return currentProductPrice > 100 ? spreadSyntaxObject(currentResult, ["expensive": spreadSyntaxArray(arrayExpensive, ["currentProduct": currentProduct] as MyObject)]) : spreadSyntaxObject(currentResult, ["cheap": spreadSyntaxArray(arrayCheap, ["currentProduct": currentProduct] as MyObject)])
}, products, ["expensive": MyArray(), "cheap": MyArray()] as MyObject)
print("grouped products: \(prettyJsonStringify(productsGrouped))")
// grouped products: {
// 	"expensive": [
// 		{
// 			"code": "pasta",
// 			"price": 321
// 		},
// 		{
// 			"code": "bubble_gum",
// 			"price": 233
// 		},
// 		{
// 			"code": "towel",
// 			"price": 499
// 		}
// 	],
// 	"cheap": [
// 		{
// 			"code": "potato_chips",
// 			"price": 5
// 		}
// 	]
// }

print("// using Swift Array.reduce() built-in method \"Array.reduce\"")

productsGrouped = products.reduce(["expensive": MyArray(), "cheap": MyArray()] as MyObject) { (currentResult: Any?, currentProduct: Any?) -> Any? in
    guard let currentResult = currentResult as? MyObject, let arrayExpensive = currentResult["expensive"], let arrayCheap = currentResult["cheap"], let currentProduct = currentProduct as? MyObject, let currentProductPrice = currentProduct["price"] as? Int else {
        return nil
    }
    return currentProductPrice > 100 ? spreadSyntaxObject(currentResult, ["expensive": spreadSyntaxArray(arrayExpensive, ["currentProduct": currentProduct] as MyObject)]) : spreadSyntaxObject(currentResult, ["cheap": spreadSyntaxArray(arrayCheap, ["currentProduct": currentProduct] as MyObject)])
}
print("grouped products: \(prettyJsonStringify(productsGrouped))")
// grouped products: {
// 	"expensive": [
// 		{
// 			"code": "pasta",
// 			"price": 321
// 		},
// 		{
// 			"code": "bubble_gum",
// 			"price": 233
// 		},
// 		{
// 			"code": "towel",
// 			"price": 499
// 		}
// 	],
// 	"cheap": [
// 		{
// 			"code": "potato_chips",
// 			"price": 5
// 		}
// 	]
// }
